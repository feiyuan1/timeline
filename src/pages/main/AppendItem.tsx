import { memo, useCallback, useMemo, useState } from 'react'
import FormModal from 'components/FormModal'
import Box from '@mui/material/Box'
import MenuItem from '@mui/material/MenuItem'
import Menu from '@mui/material/Menu'
import useToggle from 'utils/useToggle'
import { lineProps, groupProps } from '_constants/form'
import AddButton from 'components/AddButton'

const AppendItem = memo(() => {
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null)
  const open = Boolean(anchorEl)
  const [openLine, toggleLine] = useToggle()
  const [openGroup, toggleGroup] = useToggle()
  const handleOpenLine = useCallback(() => {
    handleClose()
    toggleLine()
  }, [toggleLine])

  const handleOpenGroup = useCallback(() => {
    handleClose()
    toggleGroup()
  }, [toggleGroup])

  const menuList = useMemo(
    () => [
      {
        label: '线路',
        key: 'line-add',
        clickHandler: handleOpenLine
      },
      {
        label: '线路组',
        key: 'line-group-add',
        clickHandler: handleOpenGroup
      }
    ],
    [handleOpenLine, handleOpenGroup]
  )

  const handleAdd = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget)
  }
  const handleClose = () => {
    setAnchorEl(null)
  }

  return (
    <Box>
      <AddButton onClick={handleAdd} />
      <Menu
        id="basic-menu"
        anchorEl={anchorEl}
        open={open}
        onClose={handleClose}
        MenuListProps={{
          'aria-labelledby': 'basic-button'
        }}
        anchorOrigin={{
          vertical: 'center',
          horizontal: 'left'
        }}
        transformOrigin={{ vertical: 'center', horizontal: 'left' }}
      >
        {menuList.map((item) => (
          <MenuItem onClick={item.clickHandler} key={item.key}>
            {item.label}
          </MenuItem>
        ))}
      </Menu>
      <FormModal open={openLine} handleClose={toggleLine} {...lineProps} />
      <FormModal open={openGroup} handleClose={toggleGroup} {...groupProps} />
    </Box>
  )
})

export default AppendItem
