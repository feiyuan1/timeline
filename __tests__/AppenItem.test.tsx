import {
  fireEvent,
  getAllByRole,
  getByText,
  render,
  screen
} from '@testing-library/react'
import AppendItem from 'pages/main/AppendItem'

const showMenu = (container: HTMLElement) => {
  // TODO 因为存在 test case: AddButton is visible in initial state => 这里还需要再次校验吗？
  const button = container.querySelector<HTMLButtonElement>(
    '.MuiIconButton-root'
  )!
  fireEvent.click(
    button,
    new MouseEvent('click', {
      bubbles: true,
      cancelable: true
    })
  )
  // TODO 因为存在 test case：click IconButton and Menu should be shouwn => 这里还需要再次校验吗？
  return screen.getByRole('presentation')
}

describe('AppendItem UI', () => {
  // TODO getByRole 会引发 type="submit" button 某个事件，导致表单触发提交验证
  it('AddButton is visible in initial state', () => {
    const { container } = render(<AppendItem />)
    const button = container.querySelector<HTMLButtonElement>(
      '.MuiIconButton-root'
    )
    expect(button).toBeInTheDocument()
    if (!button) {
      return
    }
    const addIcon = screen.getByTestId('AddIcon')
    expect(addIcon).toBeVisible()
    expect(button).toContainElement(addIcon)
  })

  it('Menu UI', () => {
    const { container } = render(<AppendItem />)
    const menu = showMenu(container)!
    const menuList = getAllByRole(menu, 'menuitem')
    expect(menuList.length).toBe(2)
    expect(getByText(menu, '线路')).toBeInTheDocument()
    expect(getByText(menu, '线路组')).toBeInTheDocument()
  })
  it('LineModal', () => {})
  it('LineGroupModal', () => {})
})

describe('AppendItem interaction', () => {
  it('click IconButton and Menu should be shouwn', () => {
    const { container } = render(<AppendItem />)
    const menu = showMenu(container)
    expect(menu).toHaveClass('MuiMenu-root')
  })

  it('click add line button in menu', () => {
    //  Menu should be closed; LineModal is visible
  })
  it('click add lineGroup button in menu', () => {
    //  Menu should be closed; LineGroupModal is visible
  })
})

describe('AppendItem logic', () => {
  it('Menu is hidden by default', () => {
    render(<AppendItem />)
    const menu = screen.queryByRole('presentation')
    expect(menu).toBe(null)
  })

  it('LineModal is hidden by default', () => {})
  it('LineGroupModal is hidden by default', () => {})
})
