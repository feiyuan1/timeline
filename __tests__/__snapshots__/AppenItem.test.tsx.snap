// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppendItem UI LineModal UI 1`] = `
<div
  aria-labelledby="form-modal-title"
  class="MuiModal-root css-1g04pbp-MuiModal-root"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root MuiModal-backdrop css-4nmryk-MuiBackdrop-root-MuiModal-backdrop"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-testid="sentinelStart"
    tabindex="0"
  />
  <form
    class="MuiBox-root css-1wnsr1i"
    tabindex="-1"
  >
    <h2
      class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
      id="form-modal-title"
    >
      新增线路
    </h2>
    <div
      class="MuiStack-root css-1sos3zc-MuiStack-root"
    >
      <div
        class="MuiStack-root css-nohnhc-MuiStack-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-2hy7sk-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":ra:"
            id=":ra:-label"
          >
            名称
            <span
              aria-hidden="true"
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ljffdk-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-5h82ro-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1yrc8ca-MuiInputBase-input-MuiInput-input"
              id=":ra:"
              name="name"
              required=""
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-2hy7sk-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":rb:"
            id=":rb:-label"
          >
            描述
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-5h82ro-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1yrc8ca-MuiInputBase-input-MuiInput-input"
              id=":rb:"
              name="description"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="MuiFormControl-root css-1iw3t7y-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-1ew92b2-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            id="select-label"
          >
            时间轴类型
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-j218zi-MuiInputBase-root-MuiInput-root-MuiSelect-root"
          >
            <div
              aria-controls=":rc:"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="select-label"
              class="MuiSelect-select MuiSelect-standard MuiInputBase-input MuiInput-input css-5lvf42-MuiSelect-select-MuiInputBase-input-MuiInput-input"
              role="combobox"
              tabindex="0"
            >
              时间
            </div>
            <input
              aria-hidden="true"
              aria-invalid="false"
              class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
              tabindex="-1"
              value="0"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-1gpuzre-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-filled css-18y2hez-MuiFormHelperText-root"
          />
        </div>
      </div>
      <div
        aria-orientation="vertical"
        class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical css-5vyl5u-MuiDivider-root"
        role="separator"
      />
      <div
        class="MuiBox-root css-txdpd1"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1te2h75-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="submit"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-q6abum-MuiSvgIcon-root"
            data-testid="CheckIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M9 16.17 4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
            />
          </svg>
        </button>
      </div>
    </div>
  </form>
  <div
    data-testid="sentinelEnd"
    tabindex="0"
  />
</div>
`;
